design/feature-flags.md

```diff
@@ -0,0 +1,98 @@
+the system will continue to behave exactly like it used to without breaking any
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
+or completely remove the legacy behavior.
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of "future flags" into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
+
+Using fine-grained flags will allow users of old projects to pick up specific
+new "future" behaviors by manually adding the specific keys to their `cdk.json`
+file, without risking breakage in other unexpected areas.
+
+## Alternative Considered
+ 
+We considered an alternative of "bundling" new capabilities under a single flag
+that specifies the CDK version which created the project, but this means that
+users won't have the ability to pick and choose which future capabilities they
+want to enable in case they need them but don't want to take the risk of
+unexpected changes.
+ 
+The downside of the fine-grained approach is that it could result in a "blowing
+up" new `cdk.json` files in case there will be many new breaking capabilities
+between major releases. But this is hypothetical and even if this list ends up
+with 20 features before we release the next major version, I still think the
+benefits outweigh the risks of the alternative approach.
+
+## Design
+
+Context keys for feature flags will be listed in `cx-api/lib/features.ts` and
+will take the form: `<module>:<feature>`. 
+
+For example:
+
+- `@aws-cdk/core:enableStackNameDuplicates`
+- `@aws-cdk/aws-cloudformation:doNotCapitalizeCustomResourcePropertyNames`.
+
+Using the module name will allow easy tracing of the code that consumes this
+flag.
+
+The configuration for future flags will be under `cx-api/lib/future.ts` and will
+be encoded as a simple context hash that will be injected by `cdk init` to all
+`cdk.json` files generated for new projects.
+
+We will mandate that when a feature or bug fix is introduced under a feature
+flag, the CHANGELOG will include:
+
+- The term `(future flag)` in the title.
+- A `BREAKING CHANGES` paragraph will be added which describes the *new*
+  behavior but disclaims that it will only apply to new projects created through
+  `cdk init`. It will also indicate the context key this flag uses for users who
+  wish to enable it manually in their project.
+
+## Open Issues
+
+### Discoverability
+
+How will users discover "future flags" in case they want to enable it? We can
+also point them to `cx-api/future.ts`, or we can add a CLI command that will
```

Comment:
I like this, give every possible flag a description and `cdk flags` can print for every flag:

* Name
* Effect
* Recommended value
* Current value
* Introduced at CDK version

Reply:
Let's add this after we have >5 flags, ok?